extensions/kafka-connect/src/main/java/com/hazelcast/jet/kafka/connect/impl/SourceOffsetStorageReader.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2021 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.kafka.connect.impl;
+
+import org.apache.kafka.connect.storage.OffsetStorageReader;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+class SourceOffsetStorageReader implements OffsetStorageReader {
+
+    private final Map<Map<String, ?>, Map<String, ?>> partitionsToOffset;
+
+    SourceOffsetStorageReader(Map<Map<String, ?>, Map<String, ?>> partitionsToOffset) {
+        this.partitionsToOffset = partitionsToOffset;
+    }
+
+    @Override
+    public <V> Map<String, Object> offset(Map<String, V> partition) {
+        return offsets(Collections.singletonList(partition)).get(partition);
+    }
+
+    @Override
+    public <V> Map<Map<String, V>, Map<String, Object>> offsets(Collection<Map<String, V>> partitions) {
+        Map<Map<String, V>, Map<String, Object>> map = new HashMap<>();
```

Comment:
For a second I thought my eyes/brain is deceiving me and I started to have blurry vision - those generics are awesome :P (comment not exactly to address, I know it cannot be done in other way unfortunatelly, just for fun ;) )

Reply:
Kafka wants it maps, maps everywhere

Reply:
and it's Hazelcast that is known for it's iMap, despite having many more features ;)